Python/chapter02/2.1 - Remove Dups/miguel_2.1_sol.py

```diff
@@ -0,0 +1,179 @@
+        self.head = None
+        self.tail = None
+        for num in numbers:
+            self.append_to_tail(num)
+
+    def append_to_tail(self, d: int) -> None:
+        if self.head is None:
+            self.head = Node(d)
+            self.tail = self.head
+            return
+        end = Node(d)
+        self.tail.next = end
+        self.tail = end
+
+    def append_to_head(self, d: int) -> None:
+        new_head = Node(d)
+        new_head.next = self.head
+        self.head = new_head
+
+    def __repr__(self):
+        return self.__str__()
+
+    def __str__(self):
+        if self.head is None:
+            return '<empty>'
+        ll = []
+        n = self.head
+        while n.next is not None:
+            ll.append('{} -> '.format(n.data))
+            n = n.next
+        ll.append(str(n.data))
+        return ''.join(ll)
+
+    def __eq__(self, other: object):
+        if not isinstance(other, LinkedList):
+            return NotImplemented
+        a = self.head
+        b = other.head
+        while a is not None and b is not None:
+            if a.data != b.data:
+                return False
+            # otherwise, advance both pointers
+            a = a.next
+            b = b.next
+        return a is None and b is None
+
+
+def build_linked_list(numbers: List[int]) -> LinkedList:
+    return LinkedList(*numbers)
+
+
+def remove_dups(ll: LinkedList) -> LinkedList:
+    """
+    remove_dups will remove duplicates from the
+    input linked list ll. A temporary buffer
+    is used.
+    Runtime: O(N)
+    Space Complexity: O(N)
+    :param ll: a linked list
+    :return: a linked list without duplicates
+    """
+    n = ll.head
+    unique_vals = {n.data}  # set literal
+    output_ll = LinkedList(n.data)
+    while n is not None:
+        if n.data not in unique_vals:
+            output_ll.append_to_tail(n.data)
+        unique_vals.add(n.data)
+        n = n.next
+    return output_ll
+
+
+def remove_dups_no_buffer(ll: LinkedList) -> LinkedList:
+    """
+    remove_dups will remove duplicates from the
+    input linked list ll. No temporary buffer
+    used.
+    Runtime: O(N^2)
+    Space Complexity: O(1)
+    :param ll: a linked list
+    :return: a linked list without duplicates
+    """
+    n = ll.head
+    while n is not None:
+        curr_data = n.data
+        m = n.next
+        prev = n
+        # search for duplicates from n + 1 to the
+        # end of the linked list
+        while m is not None:
+            if m.data == curr_data:
+                # re-arrange pointers to omit
+                # the duplicate
+                prev.next = m.next
+            else:
+                # otherwise, advance m and prev pointers
+                prev = m
+            m = prev.next
+        n = n.next
+    return ll
+
+
+class TestRemoveDups(unittest.TestCase):
+
+    def setUp(self):
+        self.test_cases = [
+            (
+                build_linked_list([1, 2, 3, 3, 5]),
```

Comment:
you can change all of these to use the constructor to be able to remove `build_linked_list` altogether.

```suggestion
                LinkedList(1, 2, 3, 3, 5),
```